### PR TITLE
Added Save for Later page

### LINKED ISSUE #56: New Feature : Wishlist
**Is your feature request related to a problem? Please describe.**
Wishlist or a save for later feature where one can save their favorite books for later.
Ex: When a person is randomly browsing through the website and comes across a book of his/her interest but is already preoccupied by other things and wants to keep it for later can use this feature to save the book in his/her "to read list".

**Describe the solution you'd like**
Just like web browsers have bookmark option, e-commerce websites have wish list option to have a collection of their favorite stuffs, a similar feature must be added to the Kitabe to save people's time of looking up for same book again and again.

**Are you willing to contribute to this issue?** 
Yes, I am willing to contribute to this issue as a participant in Gssoc'21.

**Describe alternatives you've considered**
I thought of adding a "recently viewed" section.

**Additional context**
As a avid book reader i always look up for new books but sometimes due to my schedule i can't read them all. Having a collection of my favorite to-read books in one section would be very helpful.

### PR DESCRIPTION
**Pull request by Fairspy**
## Description
- Added the Save for Later page
- Fixes #56 

## Affected Dependencies
NA

## How has this been tested?
All the previously, as well as newly defined tests, were passed

## GIF Before
Before

## GIF After
After

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/Praful932/Kitabe/blob/master/CONTRIBUTING.md).
- [x] My changes do not edit/add any unrequired files.
- [x] My changes are covered by tests.

### CODE REVIEW

**Comment by Fairspy** [comment]
@Praful932 @maneprajakta I have not yet worked upon the responsiveness of the library dropdown as I wanted your input on it.

File: mainapp/urls.py [hunk 1]
```diff
@@ -5,7 +5,11 @@
     path('genre_books/<genre>', views.genre_books, name='genre_books'),
     path('explore_books/', views.explore_books, name='explore_books'),
     path('book_recommendations/', views.book_recommendations, name='book_recommendations'),
-    path('library/read_books', views.read_books, name='read_books')
+    path('library/read_books', views.read_books, name='read_books'),
+    path('save/<int:id>', views.save, name='save'),
+    path('remove/<int:id>', views.remove, name='remove'),
+    path('library/to_read', views.SaveList, name='to_read')
```
Comment (maneprajakta) [inline mainapp/urls.py:11]: `library/saved_books`  would be more clear what u say @Praful932

**Review by maneprajakta — CHANGES_REQUESTED** [review]
- whether a book is already in the read book or not it should be checked in `def save` only before adding it to db `SaveList` should simply render the saved books.
- could you show how it works when no book is saved.

File: mainapp/urls.py (hunk 1 above)
Comment (Praful932) [reply mainapp/urls.py:11]: yea lets make it `saved_books`

**Review by Praful932 — CHANGES_REQUESTED** [review]
Nice work @Fairspy , some comments
- Would be good if you add checks before saving & removing
- Also make both of these views AJAX, we do not want to reload the whole thing again for the user

**Comment by Praful932** [comment]
@maneprajakta Shall we make the dropdown to 
- **Rated Books**
- **Saved for Later**
and change **Your Read** to **Recommendations** *heart* with a tooltip
past tense of read makes everything more confusing 😂
Wdyt

**Comment by maneprajakta** [comment]
@Praful932 yes!
Sounds good to me.
so, should we change the url path also to `rated_books` rather than `read_books`.

**Comment by Praful932** [comment]
Yes, so two new urls would be
- `library/rated_books` - Already rated books
- `library/saved_books` - Saved for later

And both of these come under library dropdown
We'll change the recommendations thing in another issue/ourselves

**Comment by maneprajakta** [comment]
Yes, that works!

**Comment by Fairspy** [comment]
@maneprajakta I actually wrote it in `def save` only first but then I changed it to `SaveList` as I thought if the user rated the book from the list only then the book would automatically be removed after page refresh. 
Should I still go ahead with putting it in `def save`?

This is how it looks when no book is saved:-
Behaviour

**Comment by Fairspy** [comment]
@Praful932 By checks do you mean something like an alert showing that **Are you sure you want to Add this book?**
I will work on making both of these views AJAX but it's my first time working with AJAX so it might take a while to implement it.

**Comment by Fairspy** [comment]
> 
> 
> Yes, so two new urls would be
> 
>     * `library/rated_books` - Already rated books
> 
>     * `library/saved_books` - Saved for later
> 
> 
> And both of these come under library dropdown
> We'll change the recommendations thing in another issue/ourselves

Will work on this.

**Comment by Praful932** [comment]
For checks I meant same as what @maneprajakta checking before adding, Check [this](https://github.com/Praful932/Kitabe/blob/master/mainapp/views_ajax.py#L81) out, your AJAX implementation should be very same to this with an additional method like - `remove_saved_book` , would also recommend to avoid naming views as `save`

**Comment by maneprajakta** [comment]
> @maneprajakta I actually wrote it in `def save` only first but then I changed it to `SaveList` as I thought if the user rated the book from the list only then the book would automatically be removed after page refresh.
> Should I still go ahead with putting it in `def save`? Or maybe I could put the check in `def save` as well as `SaveList` so that it checks if it's rated before saving and also removes after it's rated after it's added to the list.
> Let me know what you think.
> 
> This is how it looks when no book is saved:-
> Behaviour

@Fairspy got your point.
Putting a check on both would work.

File: mainapp/views_ajax.py [hunk 2]
```diff
@@ -99,3 +99,28 @@ def user_rate_book(request):
...
+
+def remove_saved_book(request):
+    """AJAX request when user removes book"""
+    if request.method == 'POST' and request.is_ajax():
+        bookid = request.POST.get('bookid', None)
+        if is_bookid_invalid(bookid):
+            return JsonResponse({'success': False}, status=200)
+
+        later = SaveForLater.objects.filter(user=request.user, bookid=bookid)
+        later.delete()
```
Comment (Praful932) [inline mainapp/views_ajax.py:125]:
```suggestion
        saved_book = SaveForLater.objects.filter(user=request.user, bookid=bookid)
        saved_book.delete()
```

File: templates/mainapp/layout.html [hunk 3]
```diff
@@ -138,7 +144,9 @@ <h4 class="modal-title" id="booktitle">Basic Modal</h4>
                         <div class="col-4">
                             <img class="img-book" src="" id="bookimage">
                         </div>
-
+                        {% if user.is_authenticated %}
+                        <button id = "mybutton" class = "btn">Save for Later</button>
```
Comment (Praful932) [inline templates/mainapp/layout.html:148]: Name the button something llike `save-book-button`

File: mainapp/tests.py [hunk 4]
```diff
@@ -239,3 +239,64 @@ def test_read_book_status_code(self):
...
+        )
+        self.assertEquals(response.status_code, 200)
+        self.assertIn('true', response.content.decode("utf-8"))
+        self.client.logout()
+
+    def test_after_remove(self):
+        """Test the status code of
+        remove_saved_book When a book is removed
+        """
+        book_id = 2
```
Comment (Praful932) [inline mainapp/tests.py:272]: Would be nice if you fetch the book_id dynamically so the tests don't depend on the hardcoded value here.
The first book in the dataset would also do

File: mainapp/views.py [hunk 5]
```diff
@@ -114,3 +114,24 @@ def handler500(request, *args, **argv):
     response = render(request, 'mainapp/error_handler.html')
     response.status_code = 500
     return response
+
+
+def SaveList(request):
+    "View to render to_read page"
```
Comment (Praful932) [inline mainapp/views.py:120]:
```suggestion
    """View to render Saved books page"""
```

File: mainapp/views.py [hunk 6]
```diff
@@ -114,3 +114,24 @@ def handler500(request, *args, **argv):
...
+    book_id = list(book)
+    for i in range(len(book_id)):
+        if book_id[i] in rated_books:
+            later = SaveForLater.objects.filter(user=request.user, bookid=book_id[i])
+            later.delete()
+    if len(book_id) == 0:
+        messages.info(request, 'Please Add Some Books')
+        return redirect('index')
+    books = get_book_dict(book_id)
+    num = len(books)
```
Comment (Praful932) [inline mainapp/views.py:133]:
```suggestion
    total_books = len(books)
```

**Review by Praful932 — CHANGES_REQUESTED** [review]
This is really nice work, particularly how you used the already defined modular functions to write clean code🚀
Added some comments and let's change the name `save.html` to `saved_book.html`

File: mainapp/views.py [hunk 7]
```diff
@@ -114,3 +114,24 @@ def handler500(request, *args, **argv):
...
+    book = set(SaveForLater.objects.filter(user=request.user).values_list('bookid', flat=True))
+    book_id = list(book)
+    for i in range(len(book_id)):
+        if book_id[i] in rated_books:
+            later = SaveForLater.objects.filter(user=request.user, bookid=book_id[i])
+            later.delete()
+    if len(book_id) == 0:
+        messages.info(request, 'Please Add Some Books')
+        return redirect('index')
+    books = get_book_dict(book_id)
```
Comment (maneprajakta) [inline mainapp/views.py:132]: After deleting books from saved_book which are already rated the corresponding `book_id` should also be removed from the `book_id` list.

**Review by maneprajakta — COMMENTED** [review]
Great Work :rocket:

**Comment by maneprajakta** [comment]
@Fairspy in [deployment](https://kitabe-pr-203.herokuapp.com/) can't see the button.If you could check once.

**Comment by Fairspy** [comment]
@Praful932 @maneprajakta I checked the deployment but it shows the 404 error page when I go to the saved for later page even though it works completely fine on my local machine. Is it just showing this behaviour on my end?

**Comment by Fairspy** [comment]
I checked the console it seems to be an internal server error but locally the page status is 200 only.

**Comment by Praful932** [comment]
@Fairspy Check now, should be fine. Had to run some fake migrations so the new table `SaveForLater` gets inserted into the db :)

**Comment by Praful932** [comment]
Also it would be nice if you add a validation when a book is Saved, right after clicking the button, it could turn into a bit darker version of the same with the `Saved!` then back to what it was earlier

**Review by maneprajakta — CHANGES_REQUESTED** [review]
- Bit Margin btw both 
Screenshot from 2021-04-22 12-38-11
- By default the color comes out to be black for library in small screen drop down
Screenshot from 2021-04-22 12-33-39

**Comment by maneprajakta** [comment]
@Fairspy any updates?

**Comment by Fairspy** [comment]
@maneprajakta Sorry for the delay, I actually have my mid-sem exams going on.
If you could give me time till next week to make the changes.

**Comment by Praful932** [comment]
Sure

**Comment by Fairspy** [comment]
@Praful932 @maneprajakta I have made the requested changes, please review.
Just one question I don't quite get what you mean by fetching `book_id` dynamically for tests if you could elaborate further on that.

**Comment by Praful932** [comment]
@Fairspy dynamically as in instead of hardcoding, initalize the book_id in `setUp` by accessing the book dataframe :)
`self.book_id = df.iloc[0]...`

**Comment by maneprajakta** [comment]
@Praful932 if you could deploy this PR again.

**Comment by Praful932** [comment]
@maneprajakta The save from search modal doesn't work rn due to absence of book id, Let's keep that fix for later since it would overlap with #221  rn. Wdyt

**Comment by maneprajakta** [comment]
> @maneprajakta The save from search modal doesn't work rn due to absence of book id, Let's keep that fix for later since it would overlap with #221 rn. Wdyt

Yes, it doesn't work from search because of not getting book id. Ya after #221 we can go for fixing this.

**Review by Praful932 — APPROVED** [review]
Nice work @Fairspy 🚀

**Pull request merged**